CS8785 when adding a locally built copy of Windows.Win32.winmd

## Actual behavior

After adding a reference to a custom projection metadata winmd, the source generator fails with the following warning:

```
CSC : warning CS8785: Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'An element with the same key but a different value already exists. Key: 'Windows.Win32'
```

```xml
<ItemGroup>
    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.49-beta">
        <PrivateAssets>all</PrivateAssets>
        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <ProjectionMetadataWinmd Include="C:\Sources\win32metadata\bin\Windows.Win32.winmd" />
</ItemGroup>
```

## Expected behavior

No warning / source generation to function

## Repro steps

1. Create a new baseline CsWin32-enabled C# project
2. Attempt to refer to a custom .winmd on disk
3. Observe warning/error

### Context

- CsWin32 version: 0.3.49-beta
- Win32Metadata version (if explicitly set by project): N/A
- Target Framework: net6.0
- `LangVersion` (if explicitly set by project): N/A

--- a/src/Microsoft.Windows.CsWin32/SourceGenerator.cs
+++ b/src/Microsoft.Windows.CsWin32/SourceGenerator.cs
@@ -130,10 +130,22 @@ public class SourceGenerator : ISourceGenerator
         "Configuration",
         DiagnosticSeverity.Warning,
         isEnabledByDefault: true);
+
+    public static readonly DiagnosticDescriptor NonUniqueMetadataInputs = new(
+        InputProjectionErrorId,
+        InputProjectionErrorTitle,
+        "The metadata projections input into CsWin32 must have unique names. The name \"{0}\" is used more than once.",
+        "Configuration",
+        DiagnosticSeverity.Error,
+        isEnabledByDefault: true);
 #pragma warning restore CS1591 // Missing XML comment for publicly visible type or member
+
+    private const string InputProjectionErrorId = "PInvoke010";
+    private const string InputProjectionErrorTitle = "Input projection error";
 
     private const string NativeMethodsTxtAdditionalFileName = "NativeMethods.txt";
     private const string NativeMethodsJsonAdditionalFileName = "NativeMethods.json";
+
     private static readonly char[] ZeroWhiteSpace = new char[]
     {
         '\uFEFF', // ZERO WIDTH NO-BREAK SPACE (U+FEFF)
@@ -199,7 +211,14 @@ public class SourceGenerator : ISourceGenerator
         }
 
         Docs? docs = ParseDocs(context);
-        SuperGenerator superGenerator = SuperGenerator.Combine(CollectMetadataPaths(context).Select(path => new Generator(path, docs, options, compilation, parseOptions)));
+        Generator[] generators = CollectMetadataPaths(context).Select(path => new Generator(path, docs, options, compilation, parseOptions)).ToArray();
+        if (TryFindNonUniqueValue(generators, g => g.InputAssemblyName, StringComparer.OrdinalIgnoreCase, out (Generator Item, string Value) nonUniqueGenerator))
+        {
+            context.ReportDiagnostic(Diagnostic.Create(NonUniqueMetadataInputs, null, nonUniqueGenerator.Value));
+            return;
+        }
+
+        SuperGenerator superGenerator = SuperGenerator.Combine(generators);
         try
         {
             foreach (AdditionalText nativeMethodsTxtFile in nativeMethodsTxtFiles)
@@ -344,6 +363,23 @@ public class SourceGenerator : ISourceGenerator
         sb.AppendLine(ex.ToString());
 
         return sb.ToString();
+    }
+
+    private static bool TryFindNonUniqueValue<T, TValue>(IEnumerable<T> sequence, Func<T, TValue> valueSelector, IEqualityComparer<TValue> comparer, out (T Item, TValue Value) nonUniqueValue)
+    {
+        HashSet<TValue> seenValues = new(comparer);
+        nonUniqueValue = default;
+        foreach (T item in sequence)
+        {
+            TValue value = valueSelector(item);
+            if (!seenValues.Add(value))
+            {
+                nonUniqueValue = (item, value);
+                return true;
+            }
+        }
+
+        return false;
     }
 
     private static IReadOnlyList<string> CollectMetadataPaths(GeneratorExecutionContext context)

--- a/test/Microsoft.Windows.CsWin32.Tests/CSharpSourceGeneratorVerifier.cs
+++ b/test/Microsoft.Windows.CsWin32.Tests/CSharpSourceGeneratorVerifier.cs
@@ -21,7 +21,6 @@ internal static class CSharpSourceGeneratorVerifier
 
             this.ReferenceAssemblies = MyReferenceAssemblies.NetStandard20;
             this.TestState.Sources.Add(string.Empty);
-            this.TestState.AnalyzerConfigFiles.Add(("/.globalconfig", ConstructGlobalConfigString()));
         }
 
         public LanguageVersion LanguageVersion { get; set; } = LanguageVersion.CSharp9;
@@ -31,15 +30,11 @@ internal static class CSharpSourceGeneratorVerifier
         [StringSyntax(StringSyntaxAttribute.Json)]
         public string? NativeMethodsJson { get; set; }
 
-        protected override IEnumerable<Type> GetSourceGenerators()
-        {
-            yield return typeof(SourceGenerator);
-        }
+        public GeneratorConfiguration GeneratorConfiguration { get; set; } = GeneratorConfiguration.Default;
 
-        protected override ParseOptions CreateParseOptions()
-        {
-            return ((CSharpParseOptions)base.CreateParseOptions()).WithLanguageVersion(this.LanguageVersion);
-        }
+        protected override IEnumerable<Type> GetSourceGenerators() => [typeof(SourceGenerator)];
+
+        protected override ParseOptions CreateParseOptions() => ((CSharpParseOptions)base.CreateParseOptions()).WithLanguageVersion(this.LanguageVersion);
 
         protected override CompilationOptions CreateCompilationOptions()
         {
@@ -62,26 +57,9 @@ internal static class CSharpSourceGeneratorVerifier
                 this.TestState.AdditionalFiles.Add(("NativeMethods.json", this.NativeMethodsJson));
             }
 
+            this.TestState.AnalyzerConfigFiles.Add(("/.globalconfig", this.GeneratorConfiguration.ToGlobalConfigString()));
+
             return base.RunImplAsync(cancellationToken);
-        }
-
-        private static string ConstructGlobalConfigString(bool omitDocs = false)
-        {
-            StringBuilder globalConfigBuilder = new();
-            globalConfigBuilder.AppendLine("is_global = true");
-            globalConfigBuilder.AppendLine();
-            globalConfigBuilder.AppendLine($"build_property.CsWin32InputMetadataPaths = {JoinAssemblyMetadata("ProjectionMetadataWinmd")}");
-            if (!omitDocs)
-            {
-                globalConfigBuilder.AppendLine($"build_property.CsWin32InputDocPaths = {JoinAssemblyMetadata("ProjectionDocs")}");
-            }
-
-            return globalConfigBuilder.ToString();
-
-            static string JoinAssemblyMetadata(string name)
-            {
-                return string.Join(";", typeof(GeneratorTests).Assembly.GetCustomAttributes<AssemblyMetadataAttribute>().Where(metadata => metadata.Key == name).Select(metadata => metadata.Value));
-            }
         }
     }
 }

--- a/test/Microsoft.Windows.CsWin32.Tests/GeneratorConfiguration.cs
+++ b/test/Microsoft.Windows.CsWin32.Tests/GeneratorConfiguration.cs
@@ -1,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+internal record GeneratorConfiguration
+{
+    internal static GeneratorConfiguration Default { get; } = new();
+
+    internal ImmutableArray<string> InputMetadataPaths { get; init; } = CollectAssemblyMetadata("ProjectionMetadataWinmd");
+
+    internal ImmutableArray<string> InputDocPaths { get; init; } = CollectAssemblyMetadata("ProjectionDocs");
+
+    internal string ToGlobalConfigString()
+    {
+        StringBuilder globalConfigBuilder = new();
+        globalConfigBuilder.AppendLine("is_global = true");
+        globalConfigBuilder.AppendLine();
+        AddPathsProperty("CsWin32InputMetadataPaths", this.InputMetadataPaths);
+        AddPathsProperty("CsWin32InputDocPaths", this.InputDocPaths);
+
+        return globalConfigBuilder.ToString();
+
+        void AddPathsProperty(string name, ImmutableArray<string> paths)
+        {
+            if (!paths.IsEmpty)
+            {
+                globalConfigBuilder.AppendLine($"build_property.{name} = {string.Join("|", paths)}");
+            }
+        }
+    }
+
+    private static ImmutableArray<string> CollectAssemblyMetadata(string name) => [.. typeof(GeneratorTests).Assembly.GetCustomAttributes<AssemblyMetadataAttribute>().Where(metadata => metadata.Key == name && metadata.Value is not null).Select(metadata => metadata.Value)];
+}

--- a/test/Microsoft.Windows.CsWin32.Tests/SourceGeneratorTests.cs
+++ b/test/Microsoft.Windows.CsWin32.Tests/SourceGeneratorTests.cs
@@ -69,4 +69,21 @@ public class SourceGeneratorTests
             NativeMethodsTxt = "CreateFile",
         }.RunAsync();
     }
+
+    [Fact]
+    public async Task NonUniqueWinmdProjectionNames()
+    {
+        await new VerifyCS.Test
+        {
+            NativeMethodsTxt = "CreateFile",
+            GeneratorConfiguration = GeneratorConfiguration.Default with
+            {
+                InputMetadataPaths = GeneratorConfiguration.Default.InputMetadataPaths.AddRange(GeneratorConfiguration.Default.InputMetadataPaths),
+            },
+            ExpectedDiagnostics =
+            {
+                new DiagnosticResult(SourceGenerator.NonUniqueMetadataInputs.Id, DiagnosticSeverity.Error),
+            },
+        }.RunAsync();
+    }
 }

